src/main/java/ch/njol/skript/sections/EffSecShoot.java

```diff
@@ -0,0 +1,281 @@
+		if (finalVelocity == null || finalDirection == null)
+			return null;
+
+		for (Object shooter : shooters.getArray(event)) {
+			for (EntityData<?> entityData : types.getArray(event)) {
+				Entity finalProjectile = null;
+				Vector vector;
+				if (shooter instanceof LivingEntity livingShooter) {
+					vector = finalDirection.getDirection(livingShooter.getLocation()).multiply(finalVelocity.doubleValue());
+					//noinspection rawtypes
+					Consumer afterSpawn = afterSpawn(event, entityData, livingShooter);
+					Class<? extends Entity> type = entityData.getType();
+					Location shooterLoc = livingShooter.getLocation();
+					shooterLoc.setY(shooterLoc.getY() + livingShooter.getEyeHeight() / 2);
+					boolean isProjectile = false, useWorldSpawn = false;
+					if (Fireball.class.isAssignableFrom(type)) {
+						shooterLoc = livingShooter.getEyeLocation().add(vector.clone().normalize().multiply(0.5));
+						isProjectile = true;
+						useWorldSpawn = true;
+					} else if (Projectile.class.isAssignableFrom(type)) {
+						isProjectile = true;
+						if (trigger != null && !RUNNING_PAPER) {
+							useWorldSpawn = true;
+						}
+					}
+
+					if (isProjectile) {
+						if (useWorldSpawn) {
+							if (trigger != null) {
+								//noinspection unchecked
+                                livingShooter.getWorld().spawn(
+									shooterLoc,
+									type,
+									afterSpawn
+								);
+							} else {
+								Projectile projectile = (Projectile) livingShooter.getWorld().spawn(
+									shooterLoc,
+									type
+								);
+								projectile.setShooter(livingShooter);
+								finalProjectile = projectile;
+							}
+						} else {
+							if (trigger != null) {
+								if (launchWithBukkitConsumer != null) {
+									try {
+										launchWithBukkitConsumer.invoke(livingShooter,
+											type,
+											vector,
+											afterSpawnBukkit(event, entityData, livingShooter)
+										);
+									} catch (Exception ignored) {};
+								} else {
+									//noinspection unchecked
+									livingShooter.launchProjectile(
+										(Class<? extends Projectile>) type,
+										vector,
+										afterSpawn
+									);
+								}
+							} else {
+								//noinspection unchecked
+								finalProjectile = livingShooter.launchProjectile((Class<? extends Projectile>) type);
+								set(finalProjectile, entityData);
+							}
+						}
+					} else {
+						if (trigger != null) {
+							//noinspection unchecked
+							entityData.spawn(shooterLoc, afterSpawn);
+						} else {
+							finalProjectile = entityData.spawn(shooterLoc);
+						}
+					}
+				} else {
+					vector = finalDirection.getDirection((Location) shooter).multiply(finalVelocity.doubleValue());
+					if (trigger != null) {
+						//noinspection unchecked,rawtypes
+						entityData.spawn((Location) shooter, (Consumer) afterSpawn(event, entityData, null));
+					} else {
+						finalProjectile = entityData.spawn((Location) shooter);
+					}
+				}
```

Comment:
this is pretty hard to parse because of all the nested ifs
can you rework it using guard clause-like formatting, or with helper methods? It's just a bit hard to read with 7 levels of indentation